marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -515,3 +682,300 @@ contractLifespanUpperBound contract = case contract of
+            Nothing    -> False
+
+    in all checkValidPayment listOfPayments
+
+
+{-|
+    Check that all accounts have positive balance.
+ -}
+validateBalances :: State -> Bool
+validateBalances State{..} = all (\(_, balance) -> balance > 0) (Map.toList accounts)
+
+
+{-| Ensure that 'pendingTx' contains expected payments.   -}
+validateTxOutputs :: PendingTx -> Party -> Integer -> TransactionOutput -> Bool
+validateTxOutputs pendingTx creator deposit expectedTxOutputs = case expectedTxOutputs of
+    TransactionOutput {txOutPayments, txOutState, txOutContract} ->
+        case txOutContract of
+            Close -> let
+                -- if it's a last transaction, don't expect any continuation,
+                -- everything is payed out, including initial deposit.
+                payments = Payment creator (Ada.lovelaceOf deposit) : txOutPayments
+                in validatePayments pendingTx payments
+            -- otherwise check the continuation
+            _ -> case getContinuingOutputs pendingTx of
+                    PendingTxOut
+                        { pendingTxOutType=(ScriptTxOut _ (DataScript ds))
+                        , pendingTxOutValue
+                        } : [] -> case PlutusTx.fromData ds of
+                            Just (MarloweData expectedCreator expectedState expectedContract) -> let
+                                scriptOutputValue = Ada.getLovelace (Ada.fromValue pendingTxOutValue)
+                                validContract = expectedCreator == creator
+                                    && txOutState == expectedState
+                                    && txOutContract == expectedContract
+                                outputBalance = totalBalance (accounts txOutState)
+                                outputBalanceOk = scriptOutputValue == (outputBalance + deposit)
+                                in  outputBalanceOk
+                                    && validContract
+                                    && validatePayments pendingTx txOutPayments
+                            _ -> False
+                    _ -> False
+    Error _ -> traceErrorH "Error"
+
+
+{-|
+    Marlowe Interpreter ValidatorScript generator.
+-}
+marloweValidator
+  :: PubKey -> MarloweData -> [Input] -> PendingTx -> Bool
+marloweValidator creator MarloweData{..} inputs pendingTx@PendingTx{..} = let
+    {-  Embed contract creator public key. This makes validator script unique,
+        which makes a particular contract to have a unique script address.
+        That makes it easier to watch for contract actions inside a wallet. -}
+
+    checkCreator =
+        if marloweCreator == creator then True
+        else traceErrorH "Wrong contract creator"
+
+    {-  We require Marlowe Tx to have both lower bound and upper bounds in 'SlotRange'.
+        All are inclusive.
+    -}
+    (minSlot, maxSlot) = case pendingTxValidRange of
+        Interval (LowerBound (Finite l) True) (UpperBound (Finite h) True) -> (l, h)
```

Comment:
Seems like you could do with some standard functions for going between your intervals and `SlotRange`s.